source/common/upstream/ring_hash_lb.cc

```diff
@@ -97,8 +180,40 @@ RingHashLoadBalancer::Ring::Ring(
       config ? config.value().hash_function()
              : HashFunction::Cluster_RingHashLbConfig_HashFunction_XX_HASH;
 
+  // Populate the hash ring by walking through the (host, weight) entries in the effective_weights
+  // map, and generating (replication_factor * weight) hashes for each host. This would be
+  // straightforward if we always had integer replication factors, but if the replication factor
+  // is fractional (for reasons described above), it gets slightly more complicated. We maintain
+  // running sums -- current_replicas and target_replicas -- which allows us to handle fractional
+  // replication_factors in a somewhat stable way.
+  //
+  // For example, suppose we have N hosts, each with a weight of 1, and a replication factor of 1.5
+  // hashes per host. We start the outer loop with current_replicas = 0 and target_replicas = 0.
+  //   - For the first host, we set target_replicas = 1.5. After one run of the inner loop,
+  //     current_replicas = 1. After another run, current_replicas = 2, so the inner loop ends.
+  //   - For the second host, target_replicas becomes 3.0, and current_replicas is 2 from before.
+  //     After only one run of the inner loop, current_replicas = 3, so the inner loop ends.
+  // This continues with all even-numbered hosts getting two entries in the ring, and all odd-
+  // numbered hosts getting one entry, which averages out to 1.5 hashes per host. This math still
+  // works in the extreme case where the replication_factor is less than 1 (that is, there are more
+  // weighted hosts than capacity in the ring): some hosts will get no entries in the ring.
+  //
+  // Aside from trying to maintain an average replication factor strictly within the ring size
+  // bounds, the other goal of this algorithm is to provide fair consistent hashing behavior even
+  // when non-integer replication factors are used. For example, suppose the user specifies a fixed
+  // ring size, min=max=1024, and there are initially 512 hosts in the ring, giving a replication
+  // factor of 2. If one host becomes unhealthy, the replication factor becomes ~2.004; two of the
+  // remaining 511 healthy hosts will end up with three replicas, and all the others will still have
+  // two as before.
+  //
+  // Having said all this, it's almost always vastly preferable for the user to avoid this situation
+  // and simply specify a fixed replication factor, with ring size bounds that can accommodate some
+  // variation in the number of hosts.
```

Comment:
This algorithm seems reasonable, basically you have a a fractional error that is never more than 1 that you continue to propagate. This does badly for small effective weights, e.g. <= 2, but if you're ever this low you have a ring size that is too small to be doing ring hash LB in the first place.

Tt would be a very quick exercise to verify how well it does in practice with a few lines of Python (or C++) to generate random weights and then compute the deviation from the ideal distribution. I'm sure there is an analytical solution as well :)